Source/WebCore/platform/graphics/displaylists/DisplayListResourceHeap.h

```diff
@@ -122,193 +106,54 @@ class ResourceHeap {
         return nativeImage;
     }
 
-    std::optional<SourceImage> getSourceImage(RenderingResourceIdentifier renderingResourceIdentifier, OptionSet<ReplayOption> options = { }) const
+    std::optional<SourceImage> getSourceImage(RenderingResourceIdentifier identifier, OptionSet<ReplayOption> options = { }) const
     {
-        if (auto nativeImage = getNativeImage(renderingResourceIdentifier, options))
+        if (RefPtr nativeImage = getNativeImage(identifier, options))
             return { { *nativeImage } };
 
-        if (auto imageBuffer = getImageBuffer(renderingResourceIdentifier, options))
+        if (RefPtr imageBuffer = getImageBuffer(identifier, options))
             return { { *imageBuffer } };
 
         return std::nullopt;
     }
 
-    DecomposedGlyphs* getDecomposedGlyphs(RenderingResourceIdentifier renderingResourceIdentifier) const
+    RefPtr<DecomposedGlyphs> getDecomposedGlyphs(RenderingResourceIdentifier identifier) const
     {
-        auto* renderingResource = get<RenderingResource>(renderingResourceIdentifier);
-        return dynamicDowncast<DecomposedGlyphs>(renderingResource);
+        return m_decomposedGlyphs.get(identifier);
     }
 
-    Gradient* getGradient(RenderingResourceIdentifier renderingResourceIdentifier) const
+    RefPtr<Gradient> getGradient(RenderingResourceIdentifier identifier) const
     {
-        auto* renderingResource = get<RenderingResource>(renderingResourceIdentifier);
-        return dynamicDowncast<Gradient>(renderingResource);
+        return m_gradients.get(identifier);
     }
 
-    Filter* getFilter(RenderingResourceIdentifier renderingResourceIdentifier) const
+    RefPtr<Filter> getFilter(RenderingResourceIdentifier identifier) const
     {
-        auto* renderingResource = get<RenderingResource>(renderingResourceIdentifier);
-        return dynamicDowncast<Filter>(renderingResource);
+        return m_filters.get(identifier);
     }
 
-    Font* getFont(RenderingResourceIdentifier renderingResourceIdentifier) const
+    Font* getFont(RenderingResourceIdentifier identifier) const
```

Comment:
Any reason for this one to not return a `RefPtr`?

Reply:
When writing this, I was thinking if I was the reviewer, I would not notice this..

Yeah, Skia backend uses display lists incorrectly in raster threads, and Font is not threadsaferefcounted, so adding refs will crash gtk/wpe in EWS. Naturally that use is totally incorrect, but somebody already approved the code..